fil_fungible_token/src/token/mod.rs

```diff
@@ -362,24 +378,45 @@ where
             )));
         }
 
-        // owner and operator must exist to burn from
-        let operator = expect_id(operator)?;
-        let owner = self.resolve_to_id(owner)?;
+        // owner-initiated burn
+        if operator == owner {
+            let owner = self.resolve_to_id(owner)?;
+            return self.transaction(|state, bs| {
+                // attempt to burn the requested amount
+                let new_amount = state.change_balance_by(&bs, owner, &amount.clone().neg())?;
+                // decrease total_supply
+                state.change_supply_by(&amount.neg())?;
+                Ok(new_amount)
+            });
+        }
 
-        let new_amount = self.transaction(|state, bs| {
-            if operator != owner {
-                // attempt to use allowance and return early if not enough
-                state.attempt_use_allowance(&bs, operator, owner, amount)?;
+        // instantiate the actors if required
+        let operator = match self.get_id(operator) {
+            Ok(operator) => operator,
+            Err(MessagingError::AddressNotResolved(addr)) => {
+                // if not resolved, implicit allowance zero is not permitted to burn, so return an
+                // insufficient allowance error
+                return Err(TokenStateError::InsufficientAllowance {
+                    owner: *owner,
+                    operator: addr,
+                    allowance: TokenAmount::zero(),
+                    delta: amount.clone(),
+                }
+                .into());
             }
+            Err(e) => return Err(e.into()),
+        };
+        // resolve the owner
+        let owner = self.resolve_to_id(owner)?;
+
+        self.transaction(|state, bs| {
+            state.attempt_use_allowance(&bs, operator, owner, amount)?;
```

Comment:
`attempt_use_allowance` needs to be updated to fail if current allowance is zero.